### PR TITLE
repo: Add ostree_repo_write_regfile_inline

### PR DESCRIPTION
**Pull request by cgwalters**
When working on ostree-ext and importing from tar, it's
quite inefficient and awkward for small files to end up creating
a whole `GInputStream` and `GFileInfo` and etc. for small files.

Plus the gtk-rs binding API to map from `impl Read` to Gio
https://docs.rs/gio/0.9.1/gio/struct.ReadInputStream.html
requires that the input stream is `Send` but the Rust `tar` API
isn't.

This is only 1/3 of the problem; we also need similar APIs
to directly create a symlink, and to stream large objects via
a push-based API.

### CODE REVIEW

File: src/libostree/ostree-repo-commit.c [hunk 1]
```diff
@@ -2769,6 +2769,50 @@ ostree_repo_write_content (OstreeRepo       *self,
...
+                                  guint32           gid,
+                                  guint32           mode,
+                                  GVariant         *xattrs,
+                                  const guint8*     buf,
+                                  gsize             len,
+                                  GCancellable     *cancellable,
+                                  GError          **error)
+{
+  g_autoptr(GInputStream) memin = g_memory_input_stream_new_from_data (buf, len, NULL);
+  g_autoptr(GFileInfo) finfo = _ostree_mode_uidgid_to_gfileinfo (mode, uid, gid);
```
Comment (cgwalters) [inline src/libostree/ostree-repo-commit.c:2807]: (Now obviously this still does all the allocations, but we can optimize it later, and it's mostly about having an API that doesn't take `GInputStream*` so the Rust bindings work better)

File: src/libostree/ostree-repo-commit.c [hunk 2]
```diff
@@ -2769,6 +2769,50 @@ ostree_repo_write_content (OstreeRepo       *self,
...
+ * is intended for small files where it is reasonable to buffer the entire
+ * content in memory.
+ *
+ * Unlike `ostree_repo_write_content()`, if @expected_checksum is provided,
+ * this function will not check for the presence of the object beforehand.
+ *
+ * Returns: (transfer full): Checksum (as a hex string) of the committed file
+ */
+_OSTREE_PUBLIC
+char *
```
Comment (lucab) [inline src/libostree/ostree-repo-commit.c:2794]: Out of curiosity, isn't the convention for fallible functions to return a boolean?

File: src/libostree/libostree-devel.sym [hunk 3]
```diff
@@ -22,6 +22,11 @@
    - uncomment the include in Makefile-libostree.am
 */
 
+LIBOSTREE_2020.2 {
+global:
+  ostree_repo_write_regfile_inline;
+} LIBOSTREE_2020.1;
```
[the remark this answers is not in the document]
Comment (cgwalters) [reply src/libostree/libostree-devel.sym:28]: haha yes...this is why we have code review!

File: src/libostree/ostree-repo-commit.c (hunk 2 above)
Comment (cgwalters) [reply src/libostree/ostree-repo-commit.c:2794]: A pointer value being `NULL` signals an error by GLib convention - and gobject-introspection codifies this.  We have several other functions like this.

The problem then is APIs like this https://gitlab.gnome.org/GNOME/glib/-/commit/9966fe4493455dcdfe64483a50676891a878c72b

**Comment by cgwalters** [comment]
OK, addressed comments.

**Comment by cgwalters** [comment]
```
ERROR: tests/test-core.js

=========================

(gjs:14294): Gjs-CRITICAL **: JS ERROR: SyntaxError: identifier starts immediately after numeric literal @ /home/user/ostree/tests/test-core.js:54
```

humm exciting

**Comment by cgwalters** [comment]
OK ubuntu xenial is quite old now, its `gjs` is too old to know about octals.  Will look at reworking the travis/ubuntu/debian CI soon hopefully...

**Comment by dbnicholson** [comment]
> OK ubuntu xenial is quite old now, its `gjs` is too old to know about octals. Will look at reworking the travis/ubuntu/debian CI soon hopefully...

Xenial is actually EOL this month, so it would be a good idea to use a newer release for CI regardless.

**Comment by cgwalters** [comment]
OK this one now just needs an approval.

**Review by jlebon — APPROVED** [review]
Some comments, but not blockers.

File: src/libostree/ostree-repo-commit.c [hunk 4]
```diff
@@ -2769,6 +2769,51 @@ ostree_repo_write_content (OstreeRepo       *self,
...
+ * @buf: (array length=len) (element-type guint8): File contents
+ * @cancellable: Cancellable
+ * @error: Error
+ *
+ * Synchronously create a file object from the provided content.  This API
+ * is intended for small files where it is reasonable to buffer the entire
+ * content in memory.
+ *
+ * Unlike `ostree_repo_write_content()`, if @expected_checksum is provided,
+ * this function will not check for the presence of the object beforehand.
```
[the remark this answers is not in the document]
Comment (cgwalters) [reply src/libostree/ostree-repo-commit.c:2789]: The synchronous `stat()` traffic can add up.  In some cases we end up doing it twice which just makes us look silly.

Eventually maybe we'll use `io_uring` of course.

File: tests/test-core.js [hunk 5]
```diff
@@ -49,6 +49,20 @@ let [,dirTree] = repo.write_mtree(mtree, null);
...
 print("commit => " + commit);
 
+// Test direct write APIs
+let inline_content = "default 0.0.0.0\nloopback 127.0.0.0\nlink-local 169.254.0.0\n";
+let regfile_mode = 33188; // 0o100000 | 0o644 (but in decimal so old gjs works)
+let inline_checksum = repo.write_regfile_inline(null, 0, 0, regfile_mode, null, inline_content, null);
+assertEquals(inline_checksum, "8aaa9dc13a0c5839fe4a277756798c609c53fac6fa2290314ecfef9041065873");
+let written = false;
+try {
+    repo.write_regfile_inline("8baa9dc13a0c5839fe4a277756798c609c53fac6fa2290314ecfef9041065873", 0, 0, regfile_mode, null, inline_content, null);
```
[the remark this answers is not in the document]
Comment (cgwalters) [reply tests/test-core.js:59]: OK I have another PR coming and will change that then, want to avoid respinning all the CI for this.

**Pull request merged**